### PR TITLE
[BE] add integration test for the generation script

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #745
* __->__ #741
* #740

as titled

### CODE REVIEW

File: scripts/generate/test_generate.py [hunk 1]
```diff
@@ -95,12 +95,20 @@ def test_generate(
...
+    if seed is not None:
+        torch.manual_seed(seed)
+        # PYTHONHASHSEED can be a decimal number in the range [0, 2**32 - 1]
+        os.environ["PYTHONHASHSEED"] = str(seed % 2**32)
+        torch.use_deterministic_algorithms(True)
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
+        # env var for deterministic CuBLAS
+        # https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
```
Comment (fegin) [inline scripts/generate/test_generate.py]: Why is the `set_determinism()` removed from the test?

Comment (tianyu-l) [reply scripts/generate/test_generate.py]: This call of `set_determinism()` was added before @wconstab's PR to make RNG right, which broke this line. Given how `seed` is explicitly used here, I'm not sure if the code over there would indicate (would it still be correct?). Need @XilunWu's help on understanding more.

This PR tries to restore the behavior before the "BC-breaking change" and guard on its running.

Comment (wconstab) [reply scripts/generate/test_generate.py]: why is the set_determinism change for correct RNG affecting this code?  I would hope that calling `set_determinism(seed, deterministic=True) would be equivalent to some manual stuff done here.  What is the issue?

Comment (tianyu-l) [reply scripts/generate/test_generate.py]: discussed offline; refactored to use `set_determinism`

File: scripts/generate/test_generate.py [hunk 2]
```diff
@@ -147,6 +141,8 @@ def test_generate(
         # sequences would require https://github.com/pytorch/torchtitan/pull/686
         apply_tp_minus_sp(model, world_mesh["tp"])
 
+    utils.set_determinism(world_mesh, device, seed, deterministic=(seed is not None))
```
Comment (wconstab) [inline scripts/generate/test_generate.py]: "seed is not None" is equiavlent to the old logic, but i still think its better to separate the concepts. 

it seems mandatory to use a seed for the token sampler so that each rank uses the same token.   However, it should be optional to force kernels to run in the slower 'deterministic' mode.  So it should be possible to run with a seed but not assume deterministic=True

File: tests/integration_tests.py [hunk 3]
```diff
@@ -377,14 +390,14 @@ def build_test_list():
             [
                 [
                     "--checkpoint.enable_checkpoint",
-                    "--experimental.pipeline_parallel_degree 2",
-                    "--training.enable_cpu_offload True",
-                    "--optimizer.early_step_in_backward",
+                ],
+                [
+                    # placeholder for the generation script's generate step
```
Comment (wconstab) [inline tests/integration_tests.py:394]: is this part WIP?

Comment (tianyu-l) [reply tests/integration_tests.py:394]: No, I wrote it this way because for generation, the first step is to call `run_llama_train.sh` and the second step is to call the generating script
https://github.com/pytorch/torchtitan/pull/741/files#diff-3b751e36d12b5fa68ae66727b4d8c6ef2cce12d4f2444b46fd882942c9f4a87fR441-R447

To make it less hacky I think we need some surgery to the file & classes.

**Review by fegin — APPROVED** [review]
LGTM

**Comment by tianyu-l** [comment]
encounter ghstack issues, reopening in https://github.com/pytorch/torchtitan/pull/749
